mrmustard/physics/ansatze.py

```diff
@@ -515,6 +639,387 @@ def __and__(self, other: PolyExpAnsatz) -> PolyExpAnsatz:
         return self.__class__(As, bs, cs)
 
 
+class DiffOpPolyExpAnsatz(PolyExpBase):
+    r"""
+    The ansatz of the Fock-Bargmann representation.
+
+    Represents the ansatz function:
+
+        :math:`F(z) = \sum_i [\sum_k c^{(i)}_k \partial_y^k \textrm{exp}((z,y)^T A_i (z,y) / 2 + (z,y)^T b_i)|_{y=0}]`
+
+    with ``k`` being a multi-index. The matrices :math:`A_i` and vectors :math:`b_i` are
+    parameters of the exponential terms in the ansatz, and :math:`z` is a vector of variables, and  and :math:`y` is a vector linked to the polynomial coefficients.
+    The dimension of ``z + y`` must be equal to the dimension of ``A`` and ``b``.
+
+        .. code-block::
+
+        >>> from mrmustard.physics.ansatze import DiffOpPolyExpAnsatz
+
+        >>> A = np.array([[1.0, 0.0], [0.0, 1.0]])
+        >>> b = np.array([1.0, 1.0])
+        >>> c = np.array([[1.0,2.0,3.0]])
+
+        >>> F = DiffOpPolyExpAnsatz(A, b, c)
+        >>> z = np.array([[1.0],[2.0],[3.0]])
+
+        >>> # calculate the value of the function at the three different ``z``, since z is batched.
+        >>> val = F(z)
+
+    A and b can be batched or not, but c needs to include an explicit batch dimension that matches A and b.
+    Args:
+        A: The list of square matrices :math:`A_i`
+        b: The list of vectors :math:`b_i`
+        c: The list of arrays :math:`c_i` is coefficients for the polynomial terms in the ansatz.
+        An explicit batch dimension that matched A and b has to be given for c.
+
+    """
+
+    def __init__(
+        self,
+        A: Optional[Batch[Matrix]] = None,
+        b: Optional[Batch[Vector]] = None,
+        c: Batch[Tensor | Scalar] = np.array([[1.0]]),
+        name: str = "",
+    ):
+        self.name = name
+
+        if A is None and b is None and c is not None:
+            raise ValueError("Please provide either A or b.")
+        super().__init__(mat=A, vec=b, array=c)
+
+        if self.A.shape[0] != self.c.shape[0] or self.A.shape[0] != self.b.shape[0]:
+            raise ValueError("Batch size of A,b,c must be the same.")
+
+    @property
+    def A(self) -> Batch[ComplexMatrix]:
+        r"""
+        The list of square matrices :math:`A_i`.
+        """
+        return self.mat
+
+    @property
+    def b(self) -> Batch[ComplexVector]:
+        r"""
+        The list of vectors :math:`b_i`.
+        """
+        return self.vec
+
+    @property
+    def c(self) -> Batch[ComplexTensor]:
+        r"""
+        The array of coefficients for the polynomial terms in the ansatz.
+        """
+        return self.array
+
+    @classmethod
+    def from_function(cls, fn: Callable, **kwargs: Any) -> DiffOpPolyExpAnsatz:
```

Comment:
I believe if you inherit from PolyExpAnsatz instead of PolyExpBase then you can avoid having to copy paste `A`, `b`, `c` and `from_function` which should clean things up

Reply:
I am pretty sure that I inherent from PolyExpBase, but I can be wrong.

Reply:
Hmm even so I think it still makes sense to have `DiffOpPolyExpAnsatz` inherit from `PolyExpAnsatz` in this PR to avoid this (albeit temporary) copy pasting. When the time comes to do the replacement it should still be relatively easy I believe. Lmk what you think

Reply:
I agree from a code perspective, but the math of the two ansatze are quite different, if `PolyExpAnsatz` had supported c as an array and not just a scalar. The docstrings of the two show the differences between the two when c is not just a scalar. To me it would from a math/physics perspective be weird if one inherited from the other since they are in principle quite different once c is no longer a scalar.

Reply:
As discussed offline we agreed to keep as is